src/Features/CSharp/Portable/CodeLens/CSharpCodeLensMemberFinder.cs

```diff
@@ -0,0 +1,96 @@
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeLens;
+
+[ExportLanguageService(typeof(ICodeLensMemberFinder), LanguageNames.CSharp), Shared]
+internal sealed class CSharpCodeLensMemberFinder : ICodeLensMemberFinder
+{
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public CSharpCodeLensMemberFinder()
+    {
+    }
+
+    public async Task<ImmutableArray<CodeLensMember>> GetCodeLensMembersAsync(Document document, CancellationToken cancellationToken)
+    {
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+        using var _ = ArrayBuilder<CodeLensMember>.GetInstance(out var codeLensNodes);
+        var visitor = new CSharpCodeLensVisitor(codeLensNodes);
+
+        visitor.Visit(root);
+
+        return codeLensNodes.ToImmutable();
+    }
+
+    private sealed class CSharpCodeLensVisitor : CSharpSyntaxWalker
+    {
+        private readonly ArrayBuilder<CodeLensMember> _memberBuilder;
+
+        public CSharpCodeLensVisitor(ArrayBuilder<CodeLensMember> memberBuilder)
+        {
+            _memberBuilder = memberBuilder;
+        }
+
+        public override void VisitClassDeclaration(ClassDeclarationSyntax node)
+        {
+            _memberBuilder.Add(new CodeLensMember(node, node.Identifier.Span));
+            base.VisitClassDeclaration(node);
+        }
+
+        public override void VisitInterfaceDeclaration(InterfaceDeclarationSyntax node)
+        {
+            _memberBuilder.Add(new CodeLensMember(node, node.Identifier.Span));
+            base.VisitInterfaceDeclaration(node);
+        }
+
+        public override void VisitEnumDeclaration(EnumDeclarationSyntax node)
+        {
+            _memberBuilder.Add(new CodeLensMember(node, node.Identifier.Span));
+            base.VisitEnumDeclaration(node);
+        }
+
+        public override void VisitPropertyDeclaration(PropertyDeclarationSyntax node)
+        {
+            _memberBuilder.Add(new CodeLensMember(node, node.Identifier.Span));
+            base.VisitPropertyDeclaration(node);
+        }
+
+        public override void VisitMethodDeclaration(MethodDeclarationSyntax node)
+        {
+            _memberBuilder.Add(new CodeLensMember(node, node.Identifier.Span));
+            base.VisitMethodDeclaration(node);
+        }
+
+        public override void VisitStructDeclaration(StructDeclarationSyntax node)
+        {
+            _memberBuilder.Add(new CodeLensMember(node, node.Identifier.Span));
+            base.VisitStructDeclaration(node);
+        }
+
+        public override void VisitConstructorDeclaration(ConstructorDeclarationSyntax node)
+        {
+            _memberBuilder.Add(new CodeLensMember(node, node.Identifier.Span));
+            base.VisitConstructorDeclaration(node);
```

Comment:
do we need to recurse here?  nothing intersting can be inside a constructor (also applies to the other members).

Reply:
Good catch, shouldn't need to recurse here.  Will remove locations that can't have other interesting members.